include/blmc_robots/real_finger.hpp

```diff
@@ -373,4 +402,253 @@ robot_interfaces::finger::BackendPtr create_real_finger_backend(
+        thread_->join();
+    }
+
+    int index;
+
+    static void *write(void *instance_pointer)
+    {
+        ((Logger *)(instance_pointer))->write();
+        return nullptr;
+    }
+
+    void write()
+    {
+        // std::ofstream
+        // output_file("/local_data/sjoshi/blmc_ei/workspace/src/catkin/robots/blmc_robots/demos/./log.csv");
+        std::ofstream output_file;
+        // std::cout << "Hello";
+
+        // errno = 0;
+        output_file.open("log.csv");
+
+        // std::cerr << "Error: " << strerror(errno) << std::endl;
+
+        output_file << "(Timestamp)"
+                    << ","
+                    << "(Current Index)"
+                    << ","
+                    << " [O] Angle J1 "
+                    << ","
+                    << " [O] Angle J2 "
+                    << ","
+                    << " [O] Angle J3 "
+                    << ","
+                    << " [O] Velocity J1 "
+                    << ","
+                    << " [O] Velocity J2 "
+                    << ","
+                    << " [O] Velocity J3 "
+                    << ","
+                    << " [O] Torque J1 "
+                    << ","
+                    << " [O] Torque J2 "
+                    << ","
+                    << " [O] Torque J3 "
+                    << ","
+                    << " [A] Applied J1 "
+                    << ","
+                    << " [A] Applied J2 "
+                    << ","
+                    << " [A] Applied J3 "
+                    << ","
+                    << " [A] Desired J1 "
+                    << ","
+                    << " [A] Desired J2 "
+                    << ","
+                    << " [A] Desired J3 "
+                    << ","
+                    << " Status " << std::endl;
+        output_file.close();
+
+        while (!destructor_was_called_ &&
+               !logger_data_->desired_action->wait_for_timeindex(0, 0.1))
+        // check wait_for_timeindex here !!
+        {
+        }
+
+        index = logger_data_->observation->newest_timeindex();
+
+        // std::cout << index << std::endl; //index is anywhere from 6 to 25
+        // (+/-), gen around 20
+
+        count = 0;
+        // meh used same index
+        for (long int t = 0; !destructor_was_called_; t++)
+        {
+            count++;
+
+            if (count == block_size_)
+            {
+                output_file.open("log.csv", std::ios_base::app);
+
+                auto t1 = std::chrono::high_resolution_clock::now();
+
+                for (int j = index; j < index + block_size_; j++)
+                {
+                    // std::cout.precision(dbl::max_digits10);
+                    // get returns the stored pointer
+                    output_file
+                        << std::fixed
+                        << logger_data_->observation->timestamp_s(j) << "," << j
+                        << "," << (*logger_data_->observation)[j].angle[0]
```

Comment:
Did you try copying the observation in a local variable (`Observation observation = (*logger_data_->observation)[j]`) and measure the time?  I am wondering if there is a significant difference in performance.